pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java

```diff
@@ -198,6 +235,41 @@ public IdealState apply(IdealState idealState) {
     }
   }
 
+  // Update the helix state when an old segment commits and a new one is to be started.
+  protected void updateHelixIdealState(final String realtimeTableName, final List<String> newInstances,
+      final String oldSegmentNameStr, final String newSegmentNameStr) {
+    HelixHelper.updateIdealState(_helixManager, realtimeTableName, new Function<IdealState, IdealState>() {
+      @Override
+      public IdealState apply(IdealState idealState) {
+        return updateForNewRealtimeSegment(idealState, newInstances, oldSegmentNameStr, newSegmentNameStr);
+      }
+    }, RetryPolicies.exponentialBackoffRetryPolicy(5, 500L, 2.0f));
+  }
+
+  protected static IdealState updateForNewRealtimeSegment(IdealState idealState,
+      final List<String> newInstances, final String oldSegmentNameStr, final String newSegmentNameStr) {
+    if (oldSegmentNameStr != null) {
```

Comment:
Ah I see what's happening, you're mutating the internal state of the ideal state. Would be better to do `idealState.setPartitionState`